### PR TITLE
Apaq 14 api definition

### PR DESCRIPTION
**Pull request by patrickcapaldo**
- changed API definition
- shortened requirements-dev to just direct dependencies
- hidden API key

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,4 +21,7 @@ __pypackages__/
 dmypy.json
 
 # docs
-_build/
\ No newline at end of file
+_build/
+
+# config
+research/config.py
```
Comment (james-zafar) [inline .gitignore]: Why would we need to ignore a specific Python file? If it's used somewhere in the code then it should be in the repo, if not then we should remove it

File: apaq-v1.0.yaml [hunk 2]
```diff
@@ -0,0 +1,113 @@
+openapi: 3.0.0
+
+info:
+  title: Airport Air Quality Prediction API
+  version: 1.0
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: I think this has to be a string in v3?

File: apaq-v1.0.yaml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+          schemas:
+            $ref: '#/defintions/PredictionResponse'
+          examples:
+            applications/json:
+              air_quality:
+                - 2.12
+        '400':
+          $ref: '#/responses/BadRequest'
+
+definitions:
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: `definitions` is not a supported field in OpenAPI 3.0.0. You should use `components` instead.

File: apaq-v1.0.yaml [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        content:
+          application/json:
+            schema:
+              $ref: '#/definitions/AirportData'
+      responses:
+        '201':
+          description: |
+            An air quality prediction has successfully been created and returned.
+          schemas:
+            $ref: '#/defintions/PredictionResponse'
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: This format is invalid in OpenAPI 3.0. The format should look something more like
``` yaml
responses:
  '201':
    description: ...
    content:
      application/json:
        schema:
          $ref: '#/components/...'
          example:
            application/json: ...
```

File: apaq-v1.0.yaml [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+          examples:
+            applications/json:
+              air_quality:
+                - 2.12
+        '400':
+          $ref: '#/responses/BadRequest'
+
+definitions:
+  AirportData:
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: Just the noun is sufficient for describing input components e.g. `Airport` instead of `AirportData` here

File: apaq-v1.0.yaml [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    type: object
+    properties:
+      altitude:
+        type: number
+        description: Altitude of the airport [ft]
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: It's better to use full sentences in descriptions - e.g. `altitude of the airport in feet`

File: apaq-v1.0.yaml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      total_departures:
+        type: number
+        description: Total number of departing flights in the next 24 hours.
+
+  PredictionResponse:
+    type: object
+    properties:
+      air_quality:
+        type: number
+        description: Predicted air quality value [PM10]
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: I think OpenAPI 3.0 also has a `responses` section within components that might be useful here

File: apaq-v1.0.yaml [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    required:
+      - code
+      - message
+    additionalProperties: false
+
+responses:
+  BadRequest:
+    description: |
+      The request body or parameters were incorrect
+  schema:
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: These should be in the `errors` section of `components` and perhaps it would be better to label them with the code as well?

File: apaq-v1.0.yaml [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    $ref: '#/definitions/Error'
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: Are you planning to have any sort of security or authentication on this API?

File: data/GlobalAirportDatabase.txt
Comment (james-zafar) [inline data/GlobalAirportDatabase.txt:1]: Does this need to be in the repo?

File: data/airports.json
Comment (james-zafar) [inline data/airports.json:1]: Same with this

File: data/airports_augmented.json
Comment (james-zafar) [inline data/airports_augmented.json:1]: And all of these - if these are the outputs of the script you've committed then I think these can be removed?

File: requirements-dev.txt [hunk 10]
```diff
@@ -1,3 +1,5 @@
 -r requirements.txt
+requests==2.31.0
```
Comment (james-zafar) [inline requirements-dev.txt]: This should be in either `requirements.txt` or `requirements-dev.txt`, not both

File: requirements-dev.txt [hunk 11]
```diff
@@ -1,3 +1,5 @@
...
+pandas==2.1.4
```
Comment (james-zafar) [inline requirements-dev.txt]: Unless there's a specific reason to I think it's better to not put version constraints on dev packages

File: research/open_weather_search.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+airports_iata = []
+airports_icao = []
+airports_lat = []
+airports_lon = []
+airports_pm10 = []
+
+# Open the file with specified encoding
+with open(FILE_PATH, 'r', encoding='utf-8') as file:
+    next(file) # skips header
```
Comment (james-zafar) [inline research/open_weather_search.py]: Does the CSV library not handle this for you?

File: research/open_weather_search.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+                airports_pm10.append(float(result["list"][0]["components"]["pm10"]))
+                airports_icao.append(curr_icao)
+                airports_iata.append(curr_iata)
+                airports_lat.append(curr_lat)
+                airports_lon.append(curr_lon)
+
+            except KeyError:
+                curr_pm10 = None
+        else:
+            print("Request failed with status code:", response.status_code)
```
Comment (james-zafar) [inline research/open_weather_search.py]: Could you use the logging functionality added in the previous PR and add a more meaningful message here for consistency please?

File: research/open_weather_search.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+airports_df = pd.DataFrame({
+                            "iata": airports_iata,
+                            "icao": airports_icao,
+                            "latitude": airports_lat,
+                            "longitude": airports_lon,
+                            "pm10": airports_pm10
+                            })
+print(airports_df)
```
Comment (james-zafar) [inline research/open_weather_search.py]: Why is this needed?

File: research/open_weather_search.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+airports_df.to_csv("airports_aq_large.csv", sep='\t')
```
Comment (james-zafar) [inline research/open_weather_search.py]: Is there a specific reason to use tab separators here?

**Review by james-zafar — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (patrickxcapaldo) [reply .gitignore]: the config.py has the API key in there which i had to remove from the repo

File: apaq-v1.0.yaml (hunk 8 above)
Comment (patrickxcapaldo) [reply apaq-v1.0.yaml]: I couldn't find a place for errors in here: https://spec.openapis.org/oas/v3.0.0.html#componentsObject

File: apaq-v1.0.yaml (hunk 9 above)
Comment (patrickxcapaldo) [reply apaq-v1.0.yaml]: true, an API key is probably a good idea

File: research/open_weather_search.py (hunk 13 above)
Comment (patrickxcapaldo) [reply research/open_weather_search.py]: what would a more meaningful message be for this? would i have to parse the status code and give a unique response for each code?

File: research/open_weather_search.py (hunk 12 above)
Comment (patrickxcapaldo) [reply research/open_weather_search.py]: I had an error doing this and followed this instead:

This is because of the differences between python 2 and python 3. Use the built-in function next in python 3. That is, write next(reader) instead of reader.next() in line 47. In addition, you should open the file in the text mode. So, change line 47 as i = open( input_file, 'r' ).

https://stackoverflow.com/questions/42767250/python-csv2libsvm-py-attributeerror-csv-reader-object-has-no-attribute-nex

File: apaq-v1.0.yaml [hunk 16]
```diff
@@ -0,0 +1,111 @@
+openapi: 3.0.0
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: I think the OpenAPI spec is up to 3.0.3 now? Is there a specific reason for using 3.0.0?

File: apaq-v1.0.yaml [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+info:
+  title: Airport Air Quality Prediction API
+  version: "1.0"
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: Strings should use single quotation marks in YAML

File: apaq-v1.0.yaml [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+  description: API for predicting air quality based on airport information
+
+servers:
+  - url: /v1
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: `/v1` is a path, not a url.

File: apaq-v1.0.yaml [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+          type: array
+          description: A list of available runways at the specified airport
+          items:
+            type: object
+            properties:
+              length:
+                type: number
+                description: Length of the runway in feet
+              surface:
+                type: string
```
Comment (james-zafar) [inline apaq-v1.0.yaml:52]: Could surface be an enum?

File: apaq-v1.0.yaml [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+              example:
+                air_quality: 2.12
+        '400':
+          $ref: '#/components/responses/BadRequest'
+
+components:
+  schemas:
+    Airport:
+      type: object
+      properties:
```
Comment (james-zafar) [inline apaq-v1.0.yaml:38]: It would be good to specify the required properties as well

File: research/open_weather_search.py (hunk 15 above)
Comment (patrickxcapaldo) [reply research/open_weather_search.py]: no specific reason, should I change this?

File: apaq-v1.0.yaml [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+
+    ErrorItem:
+      description: |
+        An error message.
+      type: object
+      properties:
+        code:
+
+  responses:
+    PredictionResponse:
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: I think it might be better to identify the endpoint in the name here for extra clarity, something like `PostAirQualityResponse`?

File: apaq-v1.0.yaml [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+      required:
+        - errors
+      additionalProperties: false
+
+    ErrorItem:
+      description: |
+        An error message.
+      type: object
+      properties:
+        code:
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: This is incomplete.

File: apaq-v1.0.yaml [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+      content:
+        application/json:
+          schema:
+            $ref: '#/components/schemas/PredictionResponse'
+
+    BadRequest:
+      description: |
+        The request body or parameters were incorrect
+      schema:
+        $ref: '#/components/schemas/Error'
```
Comment (james-zafar) [inline apaq-v1.0.yaml]: This is wrong, schema must be inside a `content` field

File: research/combine_aq_airports_data.py [hunk 24]
```diff
@@ -0,0 +1,69 @@
+"""
+Program to inject air quality data into the airports list from FlightRadar24
+
+IMPORTANT:
+- This file must be run AFTER the "open_weather_search.py" file
+- This file requires a configuration file to run successfully
+- The configuration file must be in the "research" directory and be named "config.yaml"
+- The configuration file must at least contain the following:
```
Comment (james-zafar) [inline research/combine_aq_airports_data.py]: Would it not be better to use command line arguments here?

File: research/combine_aq_airports_data.py [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+            “country”: str
+            “iata”: str
+            “icao”: str
+            “name”: str
+        },
+        {},
+        ...
+    ]
+}
+"""
```
Comment (james-zafar) [inline research/combine_aq_airports_data.py]: This is quite confusing to me

File: research/combine_aq_airports_data.py [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+
+import json
+import csv
+import yaml
+from yaml.loader import SafeLoader
+
+# obtain API key from config file
+with open("config.yaml", "r", encoding="utf-8") as f:
+    config = yaml.load(f, Loader=SafeLoader)
```
Comment (james-zafar) [inline research/combine_aq_airports_data.py]: You could use `yaml.safe_load` instead

File: research/combine_aq_airports_data.py [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+    reader = csv.reader(f)
+    for row in reader:
+        air_qualities.append(row[0].split("\t"))
+
+# find matching airports and combine data
+for airport_a in airports_obj["airports"]:
+    # extract airport A
+    iata_a = airport_a["iata"]
+    for airport_b in air_qualities:
+        # extract airport B
```
Comment (james-zafar) [inline research/combine_aq_airports_data.py]: Some of these comments don't seem to add much value, perhaps it's worth reviewing and streamlining them a bit?

File: research/open_weather_search.py [hunk 28]
```diff
@@ -0,0 +1,81 @@
+"""
+Testing the IQAir API to use a list of airport locations
+to search and find air quality measurements within a particular
+distance from those locations.
+
+IMPORTANT:
+- This file must be run BEFORE the "combine_aq_airports_data.py" file
+- This file requires a configuration file to run successfully
+- The configuration file must be in the "research" directory and be named "config.yaml"
+- The configuration file must at least contain the following:
```
Comment (james-zafar) [inline research/open_weather_search.py]: If this is a prerequisite, why not just run this from the other script?

**Review by james-zafar — CHANGES_REQUESTED** [review]

File: research/open_weather_search.py (hunk 15 above)
Comment (james-zafar) [reply research/open_weather_search.py]: It's up to you, I think comma separated is more common as the name suggests (Comma Separated Values), but it doesn't make much difference

File: research/combine_aq_airports_data.py (hunk 25 above)
Comment (patrickxcapaldo) [reply research/combine_aq_airports_data.py]: if the airports.json data is not inside the repo, shouldn't i tell the user what structure that data file should have so that the script runs properly?

File: apaq-v1.0.yaml (hunk 18 above)
Comment (patrickxcapaldo) [reply apaq-v1.0.yaml]: i might just get rid of this for now

**Review by james-zafar — APPROVED** [review]
Looks good, thanks
